src/MultiAgentReinforcementLearning/src/MultiAgentReinforcementLearning.jl

```diff
@@ -0,0 +1,224 @@
+        agent(PreActStage(), env)
+    end
+end
+
+function (multiagent::MultiAgentPolicy)(::PostActStage, env::AbstractEnv)
+    for agent in multiagent
+        agent(PostActStage(), env)
+    end
+end
+
+function (multiagent::MultiAgentPolicy)(::PostEpisodeStage, env::AbstractEnv)
+    for agent in multiagent
+        agent(PostEpisodeStage(), env)
+    end
+end
+
+function (hook::MultiAgentHook)(::PreEpisodeStage, multiagent::MultiAgentPolicy, env::AbstractEnv)
+    for (name, agent) in pairs(multiagent)
+        hook[name](PreEpisodeStage(), agent, env)
+    end
+end
+
+function (hook::MultiAgentHook)(::PreActStage, multiagent::MultiAgentPolicy, env::AbstractEnv)
+    for (name, agent) in pairs(multiagent)
+        hook[name](PreActStage(), agent, env)
+    end
+end
+
+function (hook::MultiAgentHook)(::PostActStage, multiagent::MultiAgentPolicy, env::AbstractEnv)
+    for (name, agent) in pairs(multiagent)
+        hook[name](PostActStage(), agent, env)
+    end
+end
+
+function (hook::MultiAgentHook)(::PostEpisodeStage, multiagent::MultiAgentPolicy, env::AbstractEnv)
+    for (name, agent) in pairs(multiagent)
+        hook[name](PostEpisodeStage(), agent, env)
+    end
+end
+
+
+function (multiagent::MultiAgentPolicy)(env::AbstractEnv)
+    return (agent(env) for agent in multiagent)
+end
+
+function _run(
+    multiagent_policy::MultiAgentPolicy,
+    env::AbstractEnv,
+    ::Sequential,
+    stop_condition,
+    hook,
+    reset_condition,
+)
+    hook(PreExperimentStage(), multiagent_policy, env)
+    multiagent_policy(PreExperimentStage(), env)
+    is_stop = false
+    while !is_stop
+        reset!(env)
+        multiagent_policy(PreEpisodeStage(), env)
+        hook(PreEpisodeStage(), multiagent_policy, env)
+
+        while !reset_condition(multiagent_policy, env) # one episode
+            for player in current_player_iterator(env)
+                policy = multiagent_policy[player] # Select appropriate policy
+
+                policy(PreActStage(), env)
+                hook(PreActStage(), policy, env)
+
+                action = policy(env)
+                env(action)
+
+                optimise!(policy)
+
+                policy(PostActStage(), env)
+                hook(PostActStage(), policy, env)
+
+                if stop_condition(policy, env)
+                    is_stop = true
+                    policy(PreActStage(), env)
+                    hook(PreActStage(), policy, env)
+                    multiagent_policy(env)  # let the policy see the last observation
+                    break
+                end
+            end
+        end # end of an episode
+
+        if is_terminated(env)
+            multiagent_policy(PostEpisodeStage(), env)  # let the policy see the last observation
+            hook(PostEpisodeStage(), multiagent_policy, env)
+        end
+    end
+    multiagent_policy(PostExperimentStage(), env)
+    hook(PostExperimentStage(), multiagent_policy, env)
+    hook
+end
+
+function _run(
```

Comment:
What is the difference between this `_run` and the original one in RLBase? Does the RLBase one not already dispatch to this?